### PR TITLE
Enable HTTP log reporting

### PR DESCRIPTION
**Pull request by Superskyyy**
This is a work in progress. 

Now reporting logs in JSON through HTTP to `http://oap/v3/logs` does work. 

But, I'm not sure if the `oap/v3/logs` endpoint is for such usage(It seems designed for fluent-bit batch reporting?). Reporting logs one by one through HTTP may not be ideal in terms of performance. I'm not sure whether the Java agent only implements gRPC reporter intentionally out of this reason. 

Please advise.

  
Signed-off-by: YihaoChen <Superskyyy@outlook.com>

### CODE REVIEW

**Comment by tom-pytel** [comment]
> But, I'm not sure if the `oap/v3/logs` endpoint is for such usage(It seems designed for fluent-bit batch reporting?). Reporting logs one by one through HTTP may not be ideal in terms of performance. I'm not sure whether the Java agent only implements gRPC reporter intentionally out of this reason.

Not sure what the INTENDED usage of that endpoint was, but the same individual send inefficiency applies to how the spans are sent currently to `/v3/segment` instead of batching to `/v3/segments`, this should be optimized at some point. The hit is not that bad though since the http protocol uses `requests.Session` which should do persistent connections as per https://docs.python-requests.org/en/master/user/advanced/:

"The Session object allows you to persist certain parameters across requests. It also persists cookies across all requests made from the Session instance, **and will use urllib3’s connection pooling. So if you’re making several requests to the same host, the underlying TCP connection will be reused**, which can result in a significant performance increase (see HTTP persistent connection)."

In any case, if it works but may not yet be optional it is a step forward.

File: skywalking/client/http.py [hunk 1]
```diff
@@ -109,3 +110,20 @@ def report(self, generator):
...
+        self.session = requests.Session()
+
+    def fork_after_in_child(self):
+        self.session.close()
+        self.session = requests.Session()
+
+    def report(self, generator):
+        for log_data in generator:
+            json_string = json_format.MessageToJson(log_data)
+            res = self.session.post(self.url_report, json=[json.loads(json_string)])
```
Comment (tom-pytel) [inline skywalking/client/http.py:128]: Actually, looking at this suggests that the `/v3/logs` endpoint can take an array of logs so a batch send could be done. Change to:
```py
def report(self, generator):
    json = [json_format.MessageToJson(log_data) for log_data in generator]
    res = self.session.post(self.url_report, json=json)
```
Your call if you want to do now.

**Review by tom-pytel — APPROVED** [review]

File: skywalking/client/http.py (hunk 1 above)
Comment (Superskyyy) [reply skywalking/client/http.py:128]: Umm, should I make a new config entry allowing the user to choose whether batch or not?

Comment (tom-pytel) [reply skywalking/client/http.py:128]: No, batch is the right way to go (assuming this endpoint does take arrays, which is what you should check).

Comment (Superskyyy) [reply skywalking/client/http.py:128]: Got it, it does take arrays.

**Comment by Humbertzhang** [comment]
So far looks good to me, looking forward to your Kafka part, thank you.

**Comment by tom-pytel** [comment]
I would say merge this first then do Kafka as a separate PR.

**Comment by kezhenxu94** [comment]
> I would say merge this first then do Kafka as a separate PR.

I agree. @Superskyyy let's do one thing at a time, in a single PR

**Comment by Superskyyy** [comment]
> > I would say merge this first then do Kafka as a separate PR.
> 
> I agree. @Superskyyy let's do one thing at a time, in a single PR

No problem, but let me add a commit on the batch reporting first. Then lets merge.

**Comment by Superskyyy** [comment]
Oops, messed up a bit.

**Comment by kezhenxu94** [comment]
> Oops, messed up a bit.

It's ok, we will squash the commits into one when merging

**Comment by kezhenxu94** [comment]
@Superskyyy ping me when it's ready to merge

**Comment by Superskyyy** [comment]
@kezhenxu94 Checks done, ready to merge.

**Comment by kezhenxu94** [comment]
> @kezhenxu94 Checks done, ready to merge.

Thank you @Superskyyy very much 🙇🏻 , excellent work!

**Pull request merged**